lib/mps_statespace.h

```diff
@@ -144,6 +155,37 @@ class MPSStateSpace {
     }
   }
 
+  // Testing only. Convert the MPS to a wavefunction under "normal" ordering.
+  // Requires: wf be allocated beforehand with bd * 2 ^ nq -1 memory.
+  static void ToWaveFunction(MPS& state, fp_type* wf) {
+    const auto bd = state.bond_dim();
+    const auto nq = state.num_qubits();
+    fp_type* raw_state = state.get();
+
+    ConstMatrixMap accum = ConstMatrixMap((Complex*)(raw_state), 2, bd);
+    ConstMatrixMap next_block = ConstMatrixMap(nullptr, 0, 0);
+    MatrixMap result2 = MatrixMap(nullptr, 0, 0);
+    auto offset = 0;
+    auto result2_size = 2;
+
+    for (unsigned i = 1; i < nq - 1; i++) {
+      offset = GetBlockOffset(state, i);
+      // use of new does not trigger any expensive operations.
+      new (&next_block)
+          ConstMatrixMap((Complex*)(raw_state + offset), bd, 2 * bd);
+      new (&result2) MatrixMap((Complex*)(wf), result2_size, 2 * bd);
+
+      // temp variable used since result2 and accum point to same memory.
+      result2 = accum * next_block;
+      result2_size *= 2;
+      new (&accum) ConstMatrixMap((Complex*)(wf), result2_size, bd);
+    }
+    offset = GetBlockOffset(state, nq - 1);
+    new (&next_block) ConstMatrixMap((Complex*)(raw_state + offset), bd, 2);
+    new (&result2) MatrixMap((Complex*)(wf), result2_size, 2);
+    result2 = accum * next_block;
```

Comment:
Is this somehow mutating the contents of `state` or `wf`?

Reply:
`result2` is a `MatrixMap` which means it can be written to (unlike `ConstMatrixMap`). This matrix multiplication is modifying the contents of `wf`. It is not modifying any values in `raw_state`.